File: turms-client-dart/lib/src/exception/response_exception.dart
/*
 * Copyright (C) 2019 The Turms Project
 * https://github.com/turms-im/turms
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'package:fixnum/fixnum.dart';

import '../model/proto/notification/turms_notification.pb.dart';

class ResponseException implements Exception {
  final Int64? requestId;
  final int code;
  final String? reason;
  final Object? cause;
  final StackTrace? stackTrace;

  ResponseException(
      {this.requestId,
      required this.code,
      this.reason,
      this.cause,
      this.stackTrace});

  ResponseException.fromNotification(TurmsNotification notification)
      : this(
            requestId:
                notification.hasRequestId() ? notification.requestId : null,
            code: notification.code,
            reason: notification.hasReason() ? notification.reason : null);

  @override
  int get hashCode =>
      requestId.hashCode ^
      code.hashCode ^
      reason.hashCode ^
      cause.hashCode ^
      stackTrace.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is ResponseException &&
          runtimeType == other.runtimeType &&
          requestId == other.requestId &&
          code == other.code &&
          reason == other.reason &&
          cause == other.cause &&
          stackTrace == other.stackTrace;

  @override
  String toString() => '$requestId:$code:$reason:$cause:$stackTrace';
}

File: turms-client-cpp/src/turms/client/exception/response_exception.cpp
#include "turms/client/exception/response_exception.h"

namespace turms::client::exception {
ResponseException::ResponseException(int code,
                                     int64_t requestId,
                                     const std::string& reason,
                                     const std::optional<std::exception>& cause)
    : std::runtime_error(reason.empty() ? "code: " + std::to_string(code)
                                        : "code: " + std::to_string(code) + ", reason: " + reason),
      code_(code),
      requestId_(requestId),
      reason_(reason),
      cause_(cause) {
}

ResponseException::ResponseException(int code,
                                     const std::string& reason,
                                     const std::optional<std::exception>& cause)
    : std::runtime_error(reason.empty() ? "code: " + std::to_string(code)
                                        : "code: " + std::to_string(code) + ", reason: " + reason),
      code_(code),
      requestId_(std::nullopt),
      reason_(reason),
      cause_(cause) {
}

ResponseException::ResponseException(const model::proto::TurmsNotification& notification)
    : ResponseException{notification.code(),
                        notification.has_request_id() ? notification.request_id() : 0,
                        notification.has_reason() ? notification.reason() : ""} {
}

auto ResponseException::requestId() const noexcept -> const std::optional<int64_t>& {
    return requestId_;
}

auto ResponseException::code() const noexcept -> int {
    return code_;
}

auto ResponseException::reason() const noexcept -> const std::string& {
    return reason_;
}

auto ResponseException::cause() const noexcept -> const std::optional<std::exception>& {
    return cause_;
}

}  // namespace turms::client::exception